### ISSUE TITLE
Suggestion: Calculate NLL Loss on neutral preference set.

### ISSUE THREAD

**Issue by linux-leo** [xfactlab/orpo#6]
Hello to everyone who worked on ORPO. Thank you so much for releasing the Paper along with the code.

For a while now I've been interested in finding the ideal optimization algorithm for the currently biggest openly available AI preference dataset: [OpenHermesPreferences](https://huggingface.co/datasets/argilla/OpenHermesPreferences)

The dataset is derived from the [OpenHermes 2.5 dataset](https://huggingface.co/datasets/teknium/OpenHermes-2.5), which is meant for standard supervised finetuning (SFT).

So an optimization algorithm derived from standard SFT for a dataset derived from a source meant for SFT looked like an ideal fit.
Testing the algorithm on that dataset therefore might show increased performance over standard DPO, just like  with the ultrafeedback datasets, but that's when I had another idea:

Looking at the Implementation and your Paper, the NLL Loss is derived from the positive (chosen) subset of the dataset, which is also part of the log odds ratio calculation, since optimizing for the negative (rejected) subset doesn't make sense.

But in the case of the OpenHermesPreferences dataset, every response has an answer from three different sources, with the best being  in the chosen column and the worst in the rejected column. This leaves a third subset to optimize on, whose quality is in between those of the other two responses, which can be called the neutral subset. (which isn't necessarily bad or worse than the chosen response at a first glance)

Calculating and Optimizing the NLL Loss on this neutral subset in addition to optimizing the log odds ratio between the chosen and rejected answer therefore lets the AI optimize on three different responses at once, while also maybe preventing overfitting on the preference data.

What do you think about this idea?

Regards,

-Leo

**Comment by jiwooya1000** [comment]
Hello @linux-leo, 
thank you for your interest in ORPO and for suggesting a new idea!

I like the idea of fully utilizing the triplet of (rejected, neutral, chosen) responses in alignment tuning with ORPO.
One thing that we could think about is, if we have a set of chosen responses that are verified to be the best generation (relatively), adapting the model to those sets might be the best choice.
It reminds me of [Zhou et al., 2023](https://arxiv.org/abs/2305.11206) and [Sharma et al., 2024](https://arxiv.org/abs/2402.12366), which have shown that a "good" SFT dataset can also significantly improve the model's alignment. 
So, on top of the insights found in those papers, I think strictly learning the chosen responses through NLL loss (what SFT does) + learning to differentiate chosen/rejected responses would lead to the best performance, which is the current formulation.

One more thing we could try with your idea is adding some hard preference pairs like (chosen, neutral) rather than (chosen, rejected) to further improve the model and prevent it from being sub-optimal.

Either way, I think both directions are definitely worth giving a try! Would love to have more discussion on this😃

**Comment by jiwooya1000** [comment]
Closing the issue as it is not active at the moment. Please open it again whenever you would like to open the discussion again🙂
